### PR TITLE
support multi architecture image build

### PR DESCRIPTION
**Pull request by gy95**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
support multi architecture image build
use Buildx to build multi arch images: this holds all arch images into one.
But we need to open docker buildx function first: this is not easy and need many steps, I follow this blog https://medium.com/@artur.klauser/building-multi-architecture-docker-images-with-buildx-27d80f7e2408 to enable buildx
And buildx will directly push built images to the registry, but not store them locally, so we need an image repo to hold the images.
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by zc2638** [comment]
Will this operation be used in cicd?

**Comment by zc2638** [comment]
If used in cicd, I'm afraid you can't use `--push` directly

**Comment by gy95** [comment]
> Will this operation be used in cicd?

no, not used in CI/CD, this is too slow

> Will this operation be used in cicd?

No, buildx costs too much and is too slow. And if we need to store multi arch images, we must specify the `--push`
blog says like this:
```
The docker buildx build subcommand has a number of flags which determine where the final image will be stored. By default, i.e. if none of the flags are specified, the resulting image will remain captive in docker’s internal build cache. This is unlike the regular docker build command which stores the resulting image in the local docker images list. The important flags are:
--load: This flag instructs docker to load the resulting image into the local docker images list. However, this currently only works for single-architecture images. If you try this with multi-architecture images you’ll get an export error:
$ docker buildx build … --load …
…
 => ERROR exporting to oci image format 0.0s
 — — — 
 > exporting to oci image format:
 — — — 
failed to solve: rpc error: code = Unknown desc = docker
 exporter does not currently support exporting manifest lists
--push: This flag tells docker to push the resulting image to a docker registry. Your image tag has to contain the proper reference to the registry and repository name. This currently is the best way to store multi-architecture images.
```

**Comment by zc2638** [comment]
We should use the `manifest` directly so that the image can be generated locally.
This step is necessary in cicd, otherwise it is impossible to verify whether the image is available or whether it can be successfully built

**Comment by gy95** [comment]
> We should use the `manifest` directly so that the image can be generated locally. This step is necessary in cicd, otherwise it is impossible to verify whether the image is available or whether it can be successfully built

we can use `make image`, but this is only used for build local arch images.
`make crossbuildimage` can be used for build multi-arch or cross-platform images.

**Comment by gy95** [comment]
fix https://github.com/kubeedge/kubeedge/issues/3518
support multi architecture images build

File: hack/make-rules/crossbuildimage.sh [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+       targets+=("${bt%%:*}")
+     done
+  fi
+
+  for arg in "${targets[@]}"; do
+    IMAGE_NAME="$(get_imagename_by_target ${arg})"
+    DOCKERFILE_PATH="$(get_dockerfile_by_target ${arg})"
+
+    set -x
+    docker buildx build --build-arg GO_LDFLAGS="${GO_LDFLAGS}" -t ${IMAGE_REPO_NAME}/${IMAGE_NAME}:${IMAGE_TAG} -f ${DOCKERFILE_PATH} --platform linux/amd64,linux/arm64,linux/arm/v7 --push .
```
Comment (fisherxu) [inline hack/make-rules/crossbuildimage.sh:85]: The binaries compiled for arm and arm64 are different, list all the archs here, so the binary will all in image?

Comment (gy95) [reply hack/make-rules/crossbuildimage.sh:85]: yes

**Comment by zhu733756** [comment]
Marked to https://github.com/docker/setup-buildx-action.  Seems that we can use this action to test buildx.

**Comment by fisherxu** [comment]
> Marked to https://github.com/docker/setup-buildx-action. Seems that we can use this action to test buildx.

With this we can copy the image to local host? @zhu733756

**Comment by zhu733756** [comment]
> > Marked to https://github.com/docker/setup-buildx-action. Seems that we can use this action to test buildx.
> 
> With this we can copy the image to local host? @zhu733756

Cannot,  but we can test the process and push the image to the public registry like dockerhub by using an account with permission. And I think this solves the problem of periodic mirror push.

**Comment by fisherxu** [comment]
> Cannot, but we can test the process and push the image to the public registry like dockerhub by using an account with permission. And I think this solves the problem of periodic mirror push.

`make image` will runs in ci only for `amd64`, which can check the image built process.
How about add this `make crossbuild` to a daily built action? So no need to run for every pr and will push the daily build image.  cc @zc2638 @gy95 @zhu733756

**Comment by fisherxu** [comment]
I'm wondering how long does it take to run `make crossbuildimage` for one components a cloudcore?

**Comment by zhu733756** [comment]
> > Cannot, but we can test the process and push the image to the public registry like dockerhub by using an account with permission. And I think this solves the problem of periodic mirror push.
> 
> `make image` will runs in ci only for `amd64`, which can check the image built process. How about add this `make crossbuild` to a daily built action? So no need to run for every pr and will push the daily build image. cc @zc2638 @gy95 @zhu733756

Agree with you. Multiple architectures can be built on a regular basis, regardless of running time.

**Comment by gy95** [comment]
Using docker/setup-buildx-action@v1 to install buildx 
Now it seems to work, only need to process the  authentication problems of `docker push images`
And we can trigger this `make crossbuildimage` action only when a new tag is created.
@fisherxu @zhu733756

**Comment by zhu733756** [comment]
> Using docker/setup-buildx-action@v1 to install buildx Now it seems to work, only need to process the authentication problems of `docker push images` And we can trigger this `make crossbuildimage` action only when a new tag is created. @fisherxu @zhu733756

Do you mean using a label to trigger, not daily built?

**Comment by gy95** [comment]
now we release a version using a new tag,
so we can trigger `buildx ci` to build multi-arch images only when a new tag is created, based on this new tag.
we can do this like below. and based on the new tag, run `make crossbuildimage` cmd

```
name: Image Release

#only triggered when a new tag is created.
on:
  create:
    tags:
      - '*'

jobs:
  buildx:
    runs-on: ubuntu-18.04
    timeout-minutes: 30
    name: Buildx to build multi-arch images
    steps:
      - uses: actions/cache@v2
        with:
          path: ~/go/pkg/mod
          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}

      - name: Checkout
        uses: actions/checkout@v2
        with:
          fetch-depth: 0

      - uses: docker/setup-buildx-action@v1
        id: buildx
        with:
          install: true

      - name: Build
        run: |
          make crossbuildimage 
```

**Comment by gy95** [comment]
> > Using docker/setup-buildx-action@v1 to install buildx Now it seems to work, only need to process the authentication problems of `docker push images` And we can trigger this `make crossbuildimage` action only when a new tag is created. @fisherxu @zhu733756
> 
> Do you mean using a label to trigger, not a daily build?

yes, this cost too much time, and IMO we don't need daily build at all, we only need to do this when release a new version, and host images to the official docker repo. so users can pull images.

**Comment by zhu733756** [comment]
Thx.  Seems good.  Why not move it out to a single temple?

**Comment by gy95** [comment]
this patch only add `make crossbuildimage` cmd,  integrate with buildx in https://github.com/kubeedge/kubeedge/pull/3538

File: Makefile [hunk 2]
```diff
@@ -323,6 +323,27 @@ image:
 	hack/make-rules/image.sh $(WHAT)
 endif
 
+define CROSS_IMAGE_HELP_INFO
+# Use Buildx multi-architecture docker images.
+#	Usage of Buildx, please ref: https://medium.com/@artur.klauser/building-multi-architecture-docker-images-with-buildx-27d80f7e2408
```
Comment (fisherxu) [inline Makefile:328]: We can remove this comment, which has been added in script.

Comment (gy95) [reply Makefile]: fixed.

**Review by fisherxu — COMMENTED** [review]
/lgtm
/approve

Merge this first, let's discuss about the CICD later.

**Pull request merged**